pkg/chaosdaemon/tasks/chaosonprocess.go

```diff
@@ -0,0 +1,252 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package tasks
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/ChaosErr"
+)
+
+var ErrCanNotAdd = errors.New("can not add")
+var ErrCanNotAssign = errors.New("can not assign")
+
+// NewChaosOnProcess init ChaosOnProcess with values can not be Assigned.
+type NewChaosOnProcess interface {
+	New(immutableValues interface{}) (ChaosOnProcess, error)
+}
+
+type AssignChaosOnProcess interface {
+	Assign(ChaosOnProcess) error
+}
+
+type TaskToProcess interface {
+	Addable
+	NewChaosOnProcess
+	AssignChaosOnProcess
+}
+
+type ChaosCanRecover interface {
+	Recover(pid PID) error
+}
+
+type ChaosOnProcess interface {
+	Inject(pid PID) error
+}
+
+// ChaosOnProcessManager is a Manager for chaos tasks base on their
+// target that marked with a PID and implement with ChaosOnProcess.
+// Since we need to :
+//	Sum tasks on same PID to one task.
+// 	Create working chaos on process called ChaosOnProcess by task info.
+// 	Assign the summed task info to the ChaosOnProcess.
+// If developers wants to use functions in ChaosOnProcessManager ,
+// their imported Task need to implement interface TaskToProcess.
+// If developers wants to totally recover task successfully
+// when no task applied on the ChaosOnProcess.
+// This ChaosOnProcess must implement ChaosCanRecover.
+// If not implement ,
+// the Recover function will return a ErrNotImplement("ChaosCanRecover") error.
+type ChaosOnProcessManager struct {
+	taskManager TaskManager
+	processMap  map[PID]ChaosOnProcess
+
+	logger logr.Logger
+}
+
+func NewChaosOnProcessManager(logger logr.Logger) ChaosOnProcessManager {
+	return ChaosOnProcessManager{
+		NewTaskManager(),
+		make(map[PID]ChaosOnProcess),
+		logger,
+	}
+}
+
+func (cm ChaosOnProcessManager) CopyTaskManager() TaskManager {
+	tm := NewTaskManager()
+	for uid, task := range cm.taskManager.TaskMap {
+		tm.TaskMap[uid] = task
+	}
+	return tm
+}
+
+func (cm ChaosOnProcessManager) CopyProcessMap() map[PID]ChaosOnProcess {
+	pm := make(map[PID]ChaosOnProcess)
+	for pid, chaosOnProcess := range cm.processMap {
+		cm.processMap[pid] = chaosOnProcess
+	}
+	return pm
+}
+
+func (cm ChaosOnProcessManager) GetWithUID(id UID) (interface{}, error) {
+	return cm.taskManager.GetWithUID(id)
+}
+
+func (cm ChaosOnProcessManager) GetWithPID(pid PID) (ChaosOnProcess, error) {
+	p, ok := cm.processMap[pid]
+	if !ok {
+		return nil, ChaosErr.NotFound("PID")
```

Comment:
I think it's better to take the PID value in the error?

Reply:
I have not do that because I think since user give the PID value as input , user should manage it by themselves.